pkg/controller/compactbackup/compact_backup_controller.go

```diff
@@ -0,0 +1,448 @@
+	if !ok {
+		return
+	}
+
+	ns := job.GetNamespace()
+	jobName := job.GetName()
+	backup := c.resolveCompactBackupFromJob(ns, job)
+	if backup == nil {
+		return
+	}
+	klog.V(4).Infof("Job %s/%s deleted through %v.", ns, jobName, utilruntime.GetCaller())
+	c.updateBackup(backup)
+}
+
+func (c *Controller) updateBackup(cur interface{}) {
+	newBackup := cur.(*v1alpha1.CompactBackup)
+	ns := newBackup.GetNamespace()
+	name := newBackup.GetName()
+
+	klog.Infof("backup object %s/%s enqueue", ns, name)
+	c.enqueueBackup(newBackup)
+}
+
+// enqueueBackup enqueues the given backup in the work queue.
+func (c *Controller) enqueueBackup(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("cound't get key for object %+v: %v", obj, err))
+		return
+	}
+	c.queue.Add(key)
+}
+
+// processNextWorkItem dequeues items, processes them, and marks them done. It enforces that the syncHandler is never
+// invoked concurrently with the same key.
+func (c *Controller) processNextWorkItem() bool {
+	metrics.ActiveWorkers.WithLabelValues(c.Name()).Add(1)
+	defer metrics.ActiveWorkers.WithLabelValues(c.Name()).Add(-1)
+
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+	if err := c.sync(key.(string)); err != nil {
+		if perrors.Find(err, controller.IsRequeueError) != nil {
+			klog.Infof("Backup: %v, still need sync: %v, requeuing", key.(string), err)
+			c.queue.AddRateLimited(key)
+		} else if perrors.Find(err, controller.IsIgnoreError) != nil {
+			klog.Infof("Backup: %v, ignore err: %v", key.(string), err)
+		} else {
+			utilruntime.HandleError(fmt.Errorf("Backup: %v, sync failed, err: %v, requeuing", key.(string), err))
+			c.queue.AddRateLimited(key)
+		}
+	} else {
+		c.queue.Forget(key)
+	}
+	return true
+}
+
+func (c *Controller) sync(key string) (err error) {
+	startTime := time.Now()
+	defer func() {
+		duration := time.Since(startTime)
+		metrics.ReconcileTime.WithLabelValues(c.Name()).Observe(duration.Seconds())
+
+		if err == nil {
+			metrics.ReconcileTotal.WithLabelValues(c.Name(), metrics.LabelSuccess).Inc()
+		} else if perrors.Find(err, controller.IsRequeueError) != nil {
+			metrics.ReconcileTotal.WithLabelValues(c.Name(), metrics.LabelRequeue).Inc()
+		} else {
+			metrics.ReconcileTotal.WithLabelValues(c.Name(), metrics.LabelError).Inc()
+			metrics.ReconcileErrors.WithLabelValues(c.Name()).Inc()
+		}
+
+		klog.V(4).Infof("Finished syncing Backup %q (%v)", key, duration)
+	}()
+
+	ns, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	klog.Infof("Backup: [%s/%s] start to sync", ns, name)
+	backup, err := c.deps.CompactBackupLister.CompactBackups(ns).Get(name)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			klog.Infof("Backup has been deleted %v", key)
+			return nil
+		}
+		klog.Infof("Backup get failed %v", err)
+		return err
+	}
+
+	//Skip
+	if backup.Status.State != "" {
+		return nil
```

Comment:
you can't just return here. if the `CompactBackup` is in prepare status, but creating job failed and updating failed status failed, how to handle that? Besides that, if create job failed, it is better to retry.

Reply:
ok, Now I will check if the job is already running and sync the job state to CR if it failed